sw/device/lib/testing/sram_ctrl_testutils.h

```diff
@@ -57,36 +20,25 @@ typedef struct sram_ctrl_data {
 } sram_ctrl_data_t;
 
 /**
- * Initialize Main SRAM Controller.
- */
-sram_ctrl_t sram_ctrl_main_init(void);
-
-/**
- * Initialize Retention SRAM Controller.
- */
-sram_ctrl_t sram_ctrl_ret_init(void);
-
-/**
- * Writes data to the beginning and end of RAM.
+ * Writes `data` at the `address` in RAM.
  */
-void sram_ctrl_write_to_ram(const sram_ctrl_t *sram,
-                            const sram_ctrl_data_t *data_in);
+void sram_ctrl_testutils_write(uintptr_t address, const sram_ctrl_data_t *data);
 
 /**
- * Reads and compares data at the beginning and end of RAM.
+ * Reads data from `address` in SRAM and compares against `expected`.
  *
- * The read data is checked for equality against the `expected` data.
+ * The data is checked for equality.
  */
-bool sram_ctrl_read_from_ram_eq(const sram_ctrl_t *sram,
-                                const sram_ctrl_data_t *expected);
+bool sram_ctrl_testutils_read_check_eq(uintptr_t address,
+                                       const sram_ctrl_data_t *expected);
 
 /**
- * Reads and compares data at the beginning and end of RAM.
+ * Reads data from `address` in SRAM and compares against `expected`.
  *
- * The read data is checked for inequality against the `expected` data.
+ * The data is checked for inequality.
  */
-bool sram_ctrl_read_from_ram_neq(const sram_ctrl_t *sram,
-                                 const sram_ctrl_data_t *expected);
+bool sram_ctrl_testutils_read_check_neq(uintptr_t address,
+                                        const sram_ctrl_data_t *expected);
```

Comment:
Why is this taken by pointer rather than by value? The ABI will pass it by pointer regardless. Since we're refactoring this, can you make this be taken by value?

Reply:
There is; it semantically does not take a pointer, but rather takes it by value. This means it cannot be passed in-line. In general, we prefer to pass things by value when it is the correct semantic.

Reply:
I disagree, logically we want to pass a pointer to the `static const` data defined outside of the function. Passing it in this way, makes the intention less ambiguous.

> This means it cannot be passed in-line.

I don't think that this potential micro optimisation has much value in this particular case.

Reply:
Given this is a request to refactor existing code I don't think this PR should be blocked on this discussion even if it does touch the relevant code.

Could you put together a small PR refactoring this code to make it pass by value @silvestrst? Then we can judge the change on its own merits without blocking other work.

Reply:
Sure, can do

Reply:
Added to #8546